storm/src/main/storm/mesos/MesosNimbus.java

```diff
@@ -577,288 +409,267 @@ public void assignSlots(Topologies topologies, Map<String, Collection<WorkerSlot
-                                         Topologies topologies,
-                                         Map<OfferID, List<LaunchTask>> toLaunch,
-                                         String topologyId,
-                                         Map<OfferID, List<WorkerSlot>> offerIDtoWorkerSlotMap,
-                                         OfferID offerId) {
-    boolean usingExistingOffer = false;
-    boolean subtractedExecutorResources = false;
-    Offer offer = offers.get(offerId);
-    List<WorkerSlot> workerSlots = offerIDtoWorkerSlotMap.get(offerId);
-
-    // Note: As of writing this comment, the workerSlots belong to same topology
-    for (WorkerSlot slot : workerSlots) {
-      TopologyDetails details = topologies.getById(topologyId);
-      String workerPrefix = "";
-      if (_conf.containsKey(MesosCommon.WORKER_NAME_PREFIX)) {
-        workerPrefix = MesosCommon.getWorkerPrefix(_conf, details);
-      }
-      TaskID taskId = TaskID.newBuilder()
-          .setValue(MesosCommon.taskId(workerPrefix + slot.getNodeId(), slot.getPort()))
-          .build();
-
-      // taskIDtoOfferMap is unnecessary
-      //    1. OfferID is usable only once - That is if we ask mesos to launch task(s) on an offerID,
-      //       mesos returns remnants as a different offer with a completely different offerID
-      //    2. If MesosNimbus is restarted, all this information is lost.
-      //    3. By not clearing taskIDtoOfferMap at the end of this method, we are only wasting memory
-      // The following if condition to check taskIDtoOfferMap is unnecessary
-      //    1. This function is invoked only once per slot.
-      //    2. If the function is invoked a second time for the same slot, then it's either because the task
-      //       was finished or killed. Either way, the old offerId we used to launch the worker is invalid
-      if ((offerId == null || offer == null) && taskIDtoOfferMap.containsKey(taskId)) {
-        offer = taskIDtoOfferMap.get(taskId);
-        if (offer != null) {
-          offerId = offer.getId();
-          usingExistingOffer = true;
+
+  public Map<String, List<TaskInfo>> getTasksToLaunch(Topologies topologies,
+                                                      Map<String, Collection<WorkerSlot>> slots,
+                                                      Map<String, OfferResources> offerResourcesPerNode) {
+    Map<String, List<TaskInfo>> tasksToLaunchPerNode = new HashMap<>();
+
+    for (String topologyId : slots.keySet()) {
+      Collection<WorkerSlot> slotList = slots.get(topologyId);
+      TopologyDetails topologyDetails = topologies.getById(topologyId);
+      Set<String> hostsWithSupervisors = new HashSet<>();
+
+      double executorCpu = MesosCommon.executorCpu(mesosStormConf);
+      double executorMem = MesosCommon.executorMem(mesosStormConf);
+
+      for (WorkerSlot slot : slotList) {
+        double workerCpu = MesosCommon.topologyWorkerCpu(mesosStormConf, topologyDetails);
```

Comment:
please move these 2 lines (setting `worker{Cpu,Mem}`) up out of this for loop, to be beside the `executor{Cpu,Mem}` lookups. i.e., these values are static during the loop.